src/frequenz/sdk/actor/_power_managing/_matryoshka.py

```diff
@@ -0,0 +1,58 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""A power manager implementation that uses the matryoshka algorithm."""
+
+import math
+
+from typing_extensions import override
+
+from ...timeseries import Power
+from ._base_classes import BaseAlgorithm, Proposal
+from ._sorted_set import SortedSet
+
+
+class Matryoshka(BaseAlgorithm):
+    """The matryoshka algorithm."""
+
+    def __init__(self) -> None:
+        """Create a new instance of the matryoshka algorithm."""
+        self._battery_buckets: dict[frozenset[int], SortedSet[Proposal]] = {}
+
+    @override
+    def handle_proposal(self, proposal: Proposal) -> Power:
+        """Handle a proposal.
```

Comment:
I feel this could also use some more elaboration how it ends up with the target power it returns

Reply:
added docs in the module level.